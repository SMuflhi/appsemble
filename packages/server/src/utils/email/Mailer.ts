import { logger } from '@appsemble/node-utils';
import { createTransport, Transporter } from 'nodemailer';
import type { Options } from 'nodemailer/lib/smtp-connection';

import type { Argv } from '../../types';
import { readAsset } from '../readAsset';
import { renderEmail } from './renderEmail';

export interface Recipient {
  /**
   * The email address of the recipient.
   */
  email: string;

  /**
   * The name of the recipient.
   */
  name?: string;
}

/**
 * A class to simplify sending emails.
 */
export class Mailer {
  transport: Transporter;

  /**
   * @param argv - The CLI arguments passed to the Appsemble server.
   */
  constructor({ smtpFrom, smtpHost, smtpPass, smtpPort, smtpSecure, smtpUser }: Argv) {
    if (smtpHost) {
      const auth = (smtpUser && smtpPass && { user: smtpUser, pass: smtpPass }) || null;
      this.transport = createTransport(
        {
          port: smtpPort || smtpSecure ? 465 : 587,
          pool: true,
          host: smtpHost,
          secure: smtpSecure,
          auth,
        } as Options,
        { from: smtpFrom },
      );
    }
  }

  /**
   * Check if the SMTP connection still works
   *
   * @throws If the SMTP connection no longer works.
   */
  async verify(): Promise<void> {
    if (!this.transport) {
      logger.warn('SMTP hasn’t been configured.');
      return;
    }
    await this.transport.verify();
  }

  /**
   * Send an email using the configured SMTP transport.
   *
   * @param to - The email
   * @param templateName - The name of the Markdown email template to send
   * @param values - A key/value pair of values to use for rendering the email.
   */
  async sendTemplateEmail(
    to: Recipient,
    templateName: string,
    values: { [key: string]: string },
  ): Promise<void> {
    const template = (await readAsset(`email/${templateName}.md`, 'utf-8')) as string;
    const { html, subject, text } = await renderEmail(template, {
      ...values,
      greeting: to.name ? `Hello ${to.name}` : 'Hello',
    });

    return this.sendEmail(to.name ? `${to.name} <${to.email}>` : to.email, subject, html, text);
  }

  /**
   * Send an email using the configured SMTP transport.
   *
   * @param to - The email address of the recipient
   * @param subject - The subject of the email
   * @param html - The HTML content of the email
   * @param text - The plain-text content of the email
   */
  async sendEmail(to: string, subject: string, html: string, text: string): Promise<void> {
    if (!this.transport) {
      logger.warn('SMTP hasn’t been configured. Not sending real email.');
    }
    logger.info(`Sending email:\nTo: ${to}\nSubject: ${subject}\n\n${text}`);
    if (this.transport) {
      await this.transport.sendMail({ html, subject, text, to });
    }
    logger.verbose('Email sent succesfully.');
  }
}
