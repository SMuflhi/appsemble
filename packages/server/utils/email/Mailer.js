import { logger } from '@appsemble/node-utils';
import nodemailer from 'nodemailer';

import renderEmail from './renderEmail';

/**
 * A class to simplify sending emails.
 */
export default class Mailer {
  /**
   * @param {Object} argv The CLI arguments passed to the Appsemble server.
   */
  constructor({ smtpFrom, smtpHost, smtpPass, smtpPort, smtpSecure, smtpUser }) {
    if (smtpHost) {
      const auth = (smtpUser && smtpPass && { user: smtpUser, pass: smtpPass }) || null;
      this.transport = nodemailer.createTransport({
        port: smtpPort || smtpSecure ? 465 : 587,
        pool: true,
        host: smtpHost,
        secure: smtpSecure,
        from: smtpFrom,
        auth,
      });
    }
  }

  async verify() {
    if (!this.transport) {
      logger.warn('SMTP hasn’t been configured.');
      return;
    }
    await this.transport.verify();
  }

  async sendEmail(to, templateName, values) {
    const { html, subject, text } = await renderEmail(templateName, {
      ...values,
      greeting: to.name ? `Hello ${to.name}` : 'Hello',
    });
    if (!this.transport) {
      logger.warn('SMTP hasn’t been configured. Not sending real email.');
    }
    const stringTo = to.name ? `${to.name} <${to.email}>` : to.email;
    logger.info(`Sending email:\nTo: ${stringTo}\nSubject: ${subject}\n\n${text}`);
    if (this.transport) {
      await this.transport.sendMail({ html, subject, text, to: stringTo });
    }
    logger.verbose('Email sent succesfully.');
  }
}
