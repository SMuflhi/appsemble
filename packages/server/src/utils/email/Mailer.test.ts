import { Transporter } from 'nodemailer';

import { Mailer } from './Mailer';

let mailer: Mailer;

beforeEach(() => {
  mailer = new Mailer({ host: '' });
});

describe('verify', () => {
  it('should succeed if no transport exists', async () => {
    expect(await mailer.verify()).toBeUndefined();
  });

  it('should succeed if the transport verification succeeds', async () => {
    mailer.transport = ({
      verify: () => Promise.resolve(true as const),
    } as Partial<Transporter>) as Transporter;
    expect(await mailer.verify()).toBeUndefined();
  });

  it('should fail if the transport verification fails', async () => {
    mailer.transport = ({
      verify: () => Promise.reject(new Error('fail')),
    } as Partial<Transporter>) as Transporter;
    await expect(mailer.verify()).rejects.toThrow(new Error('fail'));
  });
});

describe('sendEmail', () => {
  it('should send emails with a name', async () => {
    mailer.transport = ({
      sendMail: jest.fn().mockResolvedValue(null),
    } as Partial<Transporter>) as Transporter;
    await mailer.sendTemplateEmail({ email: 'test@example.com', name: 'Me' }, 'resend', {
      url: 'https://example.appsemble.app/verify?code=test',
    });
    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      to: 'Me <test@example.com>',
      subject: 'Confirm account registration',
      text: expect.any(String),
      html: expect.any(String),
      attachments: [],
    });
  });

  it('should send emails without a name', async () => {
    mailer.transport = ({
      sendMail: jest.fn().mockResolvedValue(null),
    } as Partial<Transporter>) as Transporter;
    await mailer.sendTemplateEmail({ email: 'test@example.com' }, 'resend', {
      url: 'https://example.appsemble.app/verify?code=test',
    });
    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      to: 'test@example.com',
      subject: 'Confirm account registration',
      text: expect.any(String),
      html: expect.any(String),
      attachments: [],
    });
  });

  it('should not send emails when smtp is not configured', async () => {
    expect(
      await mailer.sendTemplateEmail({ email: 'test@example.com' }, 'resend', {
        url: 'https://example.appsemble.app/verify?code=test',
      }),
    ).toBeUndefined();
  });
});
