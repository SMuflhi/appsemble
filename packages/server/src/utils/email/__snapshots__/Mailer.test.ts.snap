// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange %# for locale en 1`] = `"Email changed"`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange %# for locale en 2`] = `
"Hello,

You have changed your email address for The Appsemble Team. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards

, _The Appsemble Team_"
`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange %# for locale en 3`] = `"Email changed"`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange %# for locale en 4`] = `
"Hello John Doe,

You have changed your email address for The Appsemble Team. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards

, _The Appsemble Team_"
`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange %# for locale nl 1`] = `"E-mail gewijzigd"`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange %# for locale nl 2`] = `
"Hallo,

Je hebt je e-mailadres voor The Appsemble Team gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange %# for locale nl 3`] = `"E-mail gewijzigd"`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange %# for locale nl 4`] = `
"Beste John Doe,

Je hebt je e-mailadres voor The Appsemble Team gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail resend en should render resend %# for locale en 1`] = `"Confirm account registration"`;

exports[`sendTranslatedEmail resend en should render resend %# for locale en 2`] = `
"Hello,

Someone has requested to resend the verification key for your account. Before you can use your account, we need to verify your email address.

Please click [this link](https://example.com) to vervify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email, nothing will happen to your account.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail resend en should render resend %# for locale en 3`] = `"Confirm account registration"`;

exports[`sendTranslatedEmail resend en should render resend %# for locale en 4`] = `
"Hello John Doe,

Someone has requested to resend the verification key for your account. Before you can use your account, we need to verify your email address.

Please click [this link](https://example.com) to vervify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email, nothing will happen to your account.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail resend nl should render resend %# for locale nl 1`] = `"Accountregistratie bevestigen"`;

exports[`sendTranslatedEmail resend nl should render resend %# for locale nl 2`] = `
"Hallo,

Iemand heeft gevraagd om de verificatiesleutel voor jouw account opnieuw te verzenden. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

**Heb je deze e-mail niet aangevraagd?**

Geen zorgen! Het kan zijn dat jouw e-mailadres per ongeluk is ingevoerd. Als je deze e-mail negeert of verwijdert, gebeurt er niets met jouw account.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail resend nl should render resend %# for locale nl 3`] = `"Accountregistratie bevestigen"`;

exports[`sendTranslatedEmail resend nl should render resend %# for locale nl 4`] = `
"Beste John Doe,

Iemand heeft gevraagd om de verificatiesleutel voor jouw account opnieuw te verzenden. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

**Heb je deze e-mail niet aangevraagd?**

Geen zorgen! Het kan zijn dat jouw e-mailadres per ongeluk is ingevoerd. Als je deze e-mail negeert of verwijdert, gebeurt er niets met jouw account.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail reset en should render reset %# for locale en 1`] = `"Password reset"`;

exports[`sendTranslatedEmail reset en should render reset %# for locale en 2`] = `
"Hello,

A password reset has been requested for your account. Please click [here](https://example.com) to reset your password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail reset en should render reset %# for locale en 3`] = `"Password reset"`;

exports[`sendTranslatedEmail reset en should render reset %# for locale en 4`] = `
"Hello John Doe,

A password reset has been requested for your account. Please click [here](https://example.com) to reset your password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail reset nl should render reset %# for locale nl 1`] = `"Wachtwoordherstel"`;

exports[`sendTranslatedEmail reset nl should render reset %# for locale nl 2`] = `
"Hallo,

Er is een wachtwoordherstel aangevraagd voor jouw account. Klik [hier](https://example.com) om jouw wachtwoord opnieuw in te stellen.

**Heb je deze e-mail niet aangevraagd?**

Jouw e-mailadres is mogelijk per ongeluk ingevoerd. Deze e-mail kan veilig worden genegeerd.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail reset nl should render reset %# for locale nl 3`] = `"Wachtwoordherstel"`;

exports[`sendTranslatedEmail reset nl should render reset %# for locale nl 4`] = `
"Beste John Doe,

Er is een wachtwoordherstel aangevraagd voor jouw account. Klik [hier](https://example.com) om jouw wachtwoord opnieuw in te stellen.

**Heb je deze e-mail niet aangevraagd?**

Jouw e-mailadres is mogelijk per ongeluk ingevoerd. Deze e-mail kan veilig worden genegeerd.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail welcome en should render welcome %# for locale en 1`] = `"Welcome to The Appsemble Team"`;

exports[`sendTranslatedEmail welcome en should render welcome %# for locale en 2`] = `
"Hello,

Thank you for registering your account. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail welcome en should render welcome %# for locale en 3`] = `"Welcome to The Appsemble Team"`;

exports[`sendTranslatedEmail welcome en should render welcome %# for locale en 4`] = `
"Hello John Doe,

Thank you for registering your account. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail welcome nl should render welcome %# for locale nl 1`] = `"Welkom bij The Appsemble Team"`;

exports[`sendTranslatedEmail welcome nl should render welcome %# for locale nl 2`] = `
"Hallo,

Bedankt voor het registeren van jouw account. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_The Appsemble Team_"
`;

exports[`sendTranslatedEmail welcome nl should render welcome %# for locale nl 3`] = `"Welkom bij The Appsemble Team"`;

exports[`sendTranslatedEmail welcome nl should render welcome %# for locale nl 4`] = `
"Beste John Doe,

Bedankt voor het registeren van jouw account. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_The Appsemble Team_"
`;
